apps/aehttp/src/aehttp_dispatch_ext.erl

```diff
@@ -44,45 +42,141 @@
        ) -> {Status :: cowboy:http_status(), Headers :: list(), Body :: map()}.
 
 handle_request('GetTopBlock', _, _Context) ->
-    {ok, TopBlock} = aehttp_logic:get_top(),
-    EncodedHash = aehttp_api_parser:encode(block_hash, TopBlock),
-    EncodedHeader = aehttp_api_parser:encode(header, TopBlock),
-    {200, [], maps:put(<<"hash">>, EncodedHash, EncodedHeader)};
+    case aec_chain:top_block() of
+        Block when Block =/= undefined ->
+            case aec_blocks:height(Block) of
+                0 ->
+                    Header = aec_blocks:to_header(Block),
+                    {200, [], aec_headers:serialize_for_client(Header, key)};
+                _ ->
+                    PrevBlockHash = aec_blocks:prev_hash(Block),
+                    case aec_chain:get_block(PrevBlockHash) of
+                        {ok, PrevBlock} ->
+                            PrevBlockType = aec_blocks:type(PrevBlock),
+                            Header = aec_blocks:to_header(Block),
+                            {200, [], aec_headers:serialize_for_client(Header, PrevBlockType)};
+                        error ->
+                            {404, [], #{reason => <<"Block not found">>}}
+                    end
+            end;
+        undefined ->
+            {404, [], #{reason => <<"Block not found">>}}
+    end;
 
-handle_request('GetCurrentKeyBlock', Req, _Context) ->
-    get_block(fun() -> aec_chain:top_key_block() end, Req);
+handle_request('GetCurrentKeyBlock', _Req, _Context) ->
+    case aec_chain:top_key_block() of
```

Comment:
There seems to be exactly the same logic as in `GetTopBlock`, apart from difference in `aec_chain:top_block()` call. We may want to consider to extract this to some function.

Reply:
You're right -> future refactor 💪